src/uct/base/uct_iface.c

```diff
@@ -549,3 +550,42 @@ ucs_config_field_t uct_iface_config_table[] = {
 
   {NULL}
 };
+
+ucs_status_t uct_base_ep_am_short_iov(uct_ep_h ep, uint8_t id, const uct_iov_t *iov,
+                                      size_t iovcnt)
+{
+    size_t length = 0;
+    unsigned buffer_length;
+    void *buffer;
+    size_t iov_it;
+    void *payload;
+    ucs_status_t status;
+
+    buffer_length = ucs_max(uct_iov_total_length(iov, iovcnt), sizeof(uint64_t));
```

Comment:
The implementation is based on uct_ep_am_short.
We should pass 8b header to uct_ep_am_short(..., uint64_t header, ...).
If the message is less than 8b, than the message will be copied to the buffer, and the remainder bytes of header will be zeros.

Reply:
this is not ideal, but guess there is not better option. can you pls comment this in the code to make it more obvious?

Reply:
yes, sure

Reply:
Actually there is an alternative way. E.g. introduce uint64_t variable initialized by zero. And then copy from buffer minimum of sizeof(uint64_t) and iov total length.

`
    uint64_t header = 0;
    size_t buffer_length;
    void *buffer;
    size_t iov_it;
    size_t offset;
    size_t copy_to_hdr;
    ucs_status_t status;

    buffer_length = uct_iov_total_length(iov, iovcnt);
    if (buffer_length > UCS_ALLOCA_MAX_SIZE) {
        buffer = ucs_malloc(buffer_length, "uct_base_ep_am_short_iov buffer");
    } else {
        buffer = ucs_alloca(buffer_length);
    }

    for (iov_it = 0, offset = 0; iov_it < iovcnt; ++iov_it) {
        memcpy(UCS_PTR_BYTE_OFFSET(buffer, offset), iov[iov_it].buffer,
               iov[iov_it].length);
        offset += iov[iov_it].length;
    }

    copy_to_hdr = ucs_min(sizeof(uint64_t), buffer_length);
    memcpy(&header, buffer, copy_to_hdr);

    status = uct_ep_am_short(ep, id, header,
                             UCS_PTR_BYTE_OFFSET(buffer, copy_to_hdr),
                             buffer_length - copy_to_hdr);

    if (buffer_length > UCS_ALLOCA_MAX_SIZE) {
        ucs_free(buffer);
    }

    return status;
`

Reply:
don't we also need to adjust the data copy in the first part of the function? i.e start with offset=8 and do it only of total_length>8

Reply:
Yes we can avoid extra copy. However in this case code became very tricky. Because we should handle case when iov[i] buffer should be splitted into header and payload:
`
    uint64_t header = 0;
    size_t iov_it   = 0;
    size_t offset   = 0;
    size_t length;
    void *payload;
    ucs_status_t status;

    length  = uct_iov_total_length(iov, iovcnt);

    if (length > sizeof(header)) {
        length -= sizeof(header);
        if (length > UCS_ALLOCA_MAX_SIZE) {
            payload = ucs_malloc(length, "uct_base_ep_am_short_iov buffer");
        } else {
            payload = ucs_alloca(length);
        }

        for (; offset + iov[iov_it].length < sizeof(header); ++iov_it) {
            memcpy(UCS_PTR_BYTE_OFFSET(&header, offset),
                   iov[iov_it].buffer,
                   iov[iov_it].length);
            offset += iov[iov_it].length;
        }

        memcpy(UCS_PTR_BYTE_OFFSET(&header, offset), iov[iov_it].buffer,
               sizeof(header) - offset);

        memcpy(payload, UCS_PTR_BYTE_OFFSET(iov[iov_it].buffer, offset),
               iov[iov_it].length - offset);

        for (++iov_it; iov_it < iovcnt; ++iov_it) {
            memcpy(UCS_PTR_BYTE_OFFSET(payload, offset), iov[iov_it].buffer,
                   iov[iov_it].length);
            offset += iov[iov_it].length;
        }
    } else {
        for (; iov_it < iovcnt; ++iov_it) {
            memcpy(UCS_PTR_BYTE_OFFSET(&header, offset), iov[iov_it].buffer,
                   iov[iov_it].length);
            offset += iov[iov_it].length;
        }
        length  = 0;
        payload = NULL;
    }

    status = uct_ep_am_short(ep, id, header, payload, length);

    if (length > UCS_ALLOCA_MAX_SIZE) {
        ucs_free(payload);
    }

    return status;
`

Reply:
maybe better keep as-is, i'd expect most transports will have optimized version of this func